clippy_lints/src/swap.rs

```diff
@@ -70,9 +71,107 @@ impl<'tcx> LateLintPass<'tcx> for Swap {
+    let s1 = subpart_stack(e1);
+    let s2 = subpart_stack(e2);
+    if !eq_expr_value(cx, s1[0], s2[0]) {
+        return true;
+    }
+    for (x1, x2) in s1.iter().zip(s2.iter()) {
+        if !eq_expr_value(cx, x1, x2) {
+            match (&x1.kind, &x2.kind) {
+                (ExprKind::Field(_, _), ExprKind::Field(_, _)) => return true,
+                _ => return false,
+            }
+        }
+    }
+    false
+}
+
+fn generate_swap_warning(cx: &LateContext<'_>, e1: &Expr<'_>, e2: &Expr<'_>, span: Span, is_xor_based: bool) {
+    let mut applicability = Applicability::MachineApplicable;
+
+    if !is_swappable(cx, e1, e2) {
+        if let ExprKind::Index(lhs1, idx1) = e1.kind {
+            if let ExprKind::Index(lhs2, idx2) = e2.kind {
+                if eq_expr_value(cx, lhs1, lhs2) {
+                    let ty = cx.typeck_results().expr_ty(lhs1).peel_refs();
+
+                    if matches!(ty.kind(), ty::Slice(_))
+                        || matches!(ty.kind(), ty::Array(_, _))
+                        || is_type_diagnostic_item(cx, ty, sym::vec_type)
+                        || is_type_diagnostic_item(cx, ty, sym::vecdeque_type)
+                    {
+                        if let Some(slice) = Sugg::hir_opt(cx, lhs1) {
+                            span_lint_and_sugg(
+                                cx,
+                                MANUAL_SWAP,
+                                span,
+                                &format!("this looks like you are swapping elements of `{}` manually", slice),
+                                "try",
+                                format!(
+                                    "{}.swap({}, {})",
+                                    slice.maybe_par(),
+                                    snippet_with_applicability(cx, idx1.span, "..", &mut applicability),
+                                    snippet_with_applicability(cx, idx2.span, "..", &mut applicability),
+                                ),
+                                applicability,
+                            );
+                        }
+                    }
+                }
+            }
+        }
+        return;
+    }
+
+    let (replace, what, sugg) = if let (Some(first), Some(second)) = (Sugg::hir_opt(cx, e1), Sugg::hir_opt(cx, e2)) {
+        (
+            true,
+            format!(" `{}` and `{}`", first, second),
+            format!("std::mem::swap({}, {})", first.mut_addr(), second.mut_addr()),
+        )
+    } else {
+        (true, String::new(), String::new())
+    };
+
+    span_lint_and_then(
+        cx,
+        MANUAL_SWAP,
+        span,
+        &format!("this looks like you are swapping{} manually", what),
+        |diag| {
+            if !sugg.is_empty() {
+                diag.span_suggestion(span, "try", sugg, applicability);
+
+                if replace && !is_xor_based {
+                    diag.note("or maybe you should use `std::mem::replace`?");
+                }
```

Comment:
Why is this dependent on xor?

Reply:
Why we suggest replace? I think `mem::replace` it is equal to two first statements of `temp` version and it isn't related to `xor` version. By this logic I turn it off for `xor` version.

Reply:
That makes sense.